src/tFPDF/ProtectedPDF.php

```diff
@@ -0,0 +1,271 @@
+        $this->Out('/P '.$this->int_p_value);
+    }
+
+    /**
+     * @param array|null $permissions An array of ::PERMISSION_*
+     * @param string $userPassword The password for restricted access to the document (obeying $permissions)
+     * @param string|null $ownerPassword The password for unrestricted access to the document (ignoring $permissions)
+     */
+    public function SetProtection(array $permissions = null, $userPassword = '', $ownerPassword = null)
+    {
+        $protection = self::PROTECTION_BASE;
+        foreach((array)$permissions as $permission){
+            $protection += $permission;
+        }
+
+        if (strlen($userPassword) > self::PASSWORD_LENGTH || strlen((string)$ownerPassword) > self::PASSWORD_LENGTH) {
+            throw new \InvalidArgumentException('Passwords must be no greater than '.self::PASSWORD_LENGTH.' chars');
+        }
+
+        if ($ownerPassword === null) {
+            // Generate random ownerpassword
+            $ownerPassword = substr(uniqid(mt_rand(), true).uniqid(mt_rand(), true), 0, self::PASSWORD_LENGTH);
+        }
+
+        $this->InitializeEncryption($userPassword, $ownerPassword, $protection);
+
+        $this->bol_encrypted = true;
+    }
+
+    /**
+     * Compute encryption key
+     */
+    private function InitializeEncryption($userPassword, $ownerPassword, $protection)
+    {
+        // Pad passwords
+        $userPassword  = self::FormatEncryptionPassword($userPassword);
+        $ownerPassword = self::FormatEncryptionPassword($ownerPassword);
+
+        // Compute values
+        $this->str_o_value        = $this->GenerateOvalue($userPassword, $ownerPassword);
+        $this->str_encryption_key = $this->GenerateEncryptionKey($userPassword, $this->str_o_value, $protection);
+        $this->str_u_value        = $this->GenerateUvalue($this->str_encryption_key);
+        $this->int_p_value        = $this->GeneratePvalue($protection);
+    }
+
+    /**
+     * @param string $object_index
+     * @return string The key for the specified object
+     */
+    private function ObjectKey($object_index)
+    {
+        return substr($this->MD5to16($this->str_encryption_key.pack('VXxx', $object_index)), 0, 10);
+    }
+
+    /**
+     * Get MD5 as binary string
+     */
+    private static function MD5to16($string)
+    {
+        return pack('H*', md5($string));
+    }
+
+    /**
+     * @param string $userPassword
+     * @param string $ownerPassword
+     * @return string
+     */
+    private static function GenerateOvalue($userPassword, $ownerPassword)
+    {
+        $tmp         = self::MD5to16($ownerPassword);
+        $ownerRC4Key = substr($tmp, 0, 5);
+        return self::RC4($ownerRC4Key, $userPassword);
+    }
+
+    /**
+     * @param string $str_encryption_key
+     * @return string
+     */
+    private static function GenerateUvalue($str_encryption_key)
+    {
+        return self::RC4($str_encryption_key, self::ENCRYPTION_PADDING);
+    }
```

Comment:
Similar point here about naming convention - $encryptionKey would be more suitable.